internal/rest/resources/resources.go

```diff
@@ -44,68 +45,53 @@ var InternalEndpoints = rest.Resources{
+// - The address of the server clashes with another server.
+// - The server does not have defined resources.
+func ValidateEndpoints(extensionServers []rest.Server, coreAddress string) error {
 	allExistingEndpoints := []rest.Resources{UnixEndpoints, PublicEndpoints, InternalEndpoints}
 	existingEndpointPaths := make(map[string]bool)
+	serverAddresses := map[string]bool{coreAddress: true}
 
+	// Record the paths for all internal endpoints.
 	for _, endpoints := range allExistingEndpoints {
 		for _, e := range endpoints.Endpoints {
 			url := filepath.Join(string(endpoints.PathPrefix), e.Path)
 			existingEndpointPaths[url] = true
 		}
 	}
 
-	for _, e := range extensionServerEndpoints.Endpoints {
-		url := filepath.Join(string(extensionServerEndpoints.PathPrefix), e.Path)
-		if existingEndpointPaths[url] {
-			return fmt.Errorf("Endpoint %q conflicts with internal endpoint", url)
-		}
-	}
-
-	return nil
-}
-
-// GetAndValidateCoreEndpoints extracts all endpoints from extensionServers that should be allocated to the core listener.
-// It also performs the following validations:
-// 1. Only one core API server is allowed.
-// 2. Server configurations are properly set.
-// 3. Path prefixes for endpoints belonging to a single server are not duplicated
-// 4. Enpoints defined in extensionServers do not conflict with internal endpoints.
-func GetAndValidateCoreEndpoints(extensionServers []rest.Server) ([]rest.Resources, error) {
-	var coreEndpoints []rest.Resources
-	var numCoreAPIs int
-
-	for _, extensionServer := range extensionServers {
-		if !extensionServer.CoreAPI {
-			continue
+	for _, server := range extensionServers {
+		// Ensure all servers have resources.
+		if len(server.Resources) == 0 {
+			return fmt.Errorf("Server must have defined resources")
 		}
 
-		numCoreAPIs++
-		if numCoreAPIs > 1 {
-			return nil, fmt.Errorf("Only one core API server is allowed")
-		}
+		// Ensure all servers with a defined address are unique.
+		if server.Address != (types.AddrPort{}) {
+			if serverAddresses[server.Address.String()] {
+				return fmt.Errorf("Server address %q conflicts with another endpoint", server.Address.String())
+			}
 
-		err := extensionServer.ValidateServerConfigs()
-		if err != nil {
-			return nil, err
+			serverAddresses[server.Address.String()] = true
+		} else if server.Protocol != "" {
+			return fmt.Errorf("Server protocol defined without address")
 		}
 
-		seen := make(map[string]bool)
-		for _, endpoints := range extensionServer.Resources {
-			if seen[string(endpoints.PathPrefix)] {
-				return nil, fmt.Errorf("Path prefix %q is duplicated in server configuration", endpoints.PathPrefix)
-			}
+		// Ensure no endpoint path conflicts with an internal or external endpoint.
+		for _, resource := range server.Resources {
+			for _, endpoint := range resource.Endpoints {
+				url := filepath.Join(string(resource.PathPrefix), endpoint.Path)
+				if existingEndpointPaths[url] {
```

Comment:
I think using `existingEndpointPaths` here would prevent different servers to have the same endpoints. A separate map should be created before we iterate resources of a single server to keep track of duplicated endpoints local to that server. We will still need to check against `existingEndpointPaths` for internal endpoint conflicts.

Reply:
In site manager, we have a `GET /1.0/sites` endpoint on the core listener which allows the UI to get a list of sites for display. The creation of sites happens on the site control listener with the endpoint `POST /1.0/sites` (a request from LXD would reach this endpoint). The url path would be the same across these different network listeners.

Reply:
Alright, so we can allow duplicate paths but we have to verify once we assign addresses that no `Server`s with the same address have duplicate paths. 

Because otherwise you'll get into a situation with 2 servers (`ServerA` and `ServerB`) that were supplied the same address and thus will be merged into one listener, but have different implementations of `/1.0/sites`.

Reply:
The validation now works like this:
* We enforce every `Server` must have a unique `Address`, or have no address.
  * If a server has a unique address, resource paths must be unique within that server.
  * If a server has no address (or if `ServeUnix=true`), resource paths must be unique across all servers with the same configuration, or with the internal paths.